tests/Functional/ControllerTest.php

```diff
@@ -0,0 +1,211 @@
+        self::assertJsonStringEqualsJsonString(<<<JSON
+            [
+                {"from":"Jenny"},
+                {"from":"Forest"}
+            ]
+       JSON, $response->getContent());
+    }
+
+    public function testReconstitutesOneFromRequest(): void
+    {
+        $submitted = <<<JSON
+            {"from":"Jenny"}
+        JSON;
+
+        $request = Request::create('/reconstitutesOne', Request::METHOD_POST, [], [], [], [], $submitted);
+        $request->headers->set('Content-Type', MimeTypes::APPLICATION_JSON);
+        $request->headers->set('Accept', MimeTypes::APPLICATION_JSON);
+
+        $response = $this->app->handle($request);
+
+        self::assertJsonStringEqualsJsonString($submitted, $response->getContent());
+    }
+
+    public function testReconstitutesMultipleFromRequest(): void
+    {
+        $submitted = <<<JSON
+            [
+                {"from":"Jenny"},
+                {"from":"Forest"}
+            ]
+        JSON;
+
+        $request = Request::create('reconstitutesMany', Request::METHOD_POST, [], [], [], [], $submitted);
+        $request->headers->set('Content-Type', MimeTypes::APPLICATION_JSON);
+        $request->headers->set('Accept', MimeTypes::APPLICATION_JSON);
+
+        $response = $this->app->handle($request);
+
+        self::assertJsonStringEqualsJsonString($submitted, $response->getContent());
+    }
+
+    private function deleteTempDir()
+    {
+        if (!file_exists($dir = sys_get_temp_dir() . '/' . Kernel::VERSION . '/CrudControllerTestKernel')) {
+            return;
+        }
+
+        $fs = new Filesystem();
+        $fs->remove($dir);
+    }
+}
+
+/** @internal  */
+final class TestKernel extends Kernel
+{
+    use MicroKernelTrait;
+
+
+    public function __construct()
+    {
+        parent::__construct('test', false);
+    }
+
+    public function registerBundles(): array
+    {
+        return [
+            new FrameworkBundle(),
+            new ViolinesRestBundle(),
+        ];
+    }
+
+    protected function configureRoutes(RouteCollectionBuilder $routes): void
+    {
+        $routes->addRoute(new Route('/returnsOne', [
+            '_controller' => HugController::class . '::returnsOne',
+        ]));
+
+        $routes->addRoute(new Route('/returnsMany', [
+            '_controller' => HugController::class . '::returnsMany',
+        ]));
+
+        $routes->addRoute(new Route('/reconstitutesOne', [
+            '_controller' => HugController::class . '::reconstitutesOne',
+        ]));
+
+        $routes->addRoute(new Route('reconstitutesMany', [
+            '_controller' => HugController::class . '::reconstitutesMany',
+        ]));
+    }
+
+    protected function configureContainer(ContainerBuilder $container, LoaderInterface $loader): void
+    {
+    }
+
+    public function getCacheDir(): string
+    {
+        return sys_get_temp_dir() . '/' . Kernel::VERSION . '/CrudControllerTestKernel/cache/' . $this->environment;
```

Comment:
I'd suggest:
`sys_get_temp_dir() . DIRECTORY_SEPARATOR . Kernel::VERSION . DIRECTORY_SEPARATOR . get_class($this) . DIRECTORY_SEPARATOR . cache . DIRECTORY_SEPARATOR . $this->environment;`

Reply:
Is `DIRECTORY_SEPARATOR` still needed nowaday? Havn't used it in ages. `easy-admin-bundle` does not us it either.
Did you run into any issues?

Reply:
I think you are right: https://stackoverflow.com/questions/6619020/how-can-i-define-the-directory-separator-for-both-windows-and-linux-platforms

'/' is even more readable!

Reply:
but wdyt about using `get_class($this)`